### PR TITLE
[spmd_types] fix MoE sequence-sharded combine

### PR DESCRIPTION
**Pull request by pianpwk**
Fix spmd_types SP-padding support, by requiring evenly sharded seq-dim for MoE input activations, calculating post-combine output size based on local tensor shapes.

Raises in trainer for MoE + uneven seq-dim sharding.

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #3740
* #3690
* #3673
* #3655
* #3741
* __->__ #3664

### CODE REVIEW

File: torchtitan/models/common/moe.py [hunk 1]
```diff
@@ -350,11 +350,13 @@ class Config(Module.Config):
         router: TokenChoiceTopKRouter.Config
         load_balance_coeff: float | None = 1e-3
         shared_experts: FeedForward.Config | None = None
+        enable_sp_for_dense_activations: bool = True
 
     def __init__(self, config: Config):
         super().__init__()
 
         num_experts = config.num_experts
+        self.enable_sp_for_dense_activations = config.enable_sp_for_dense_activations
```
Comment (pianpwk) [inline torchtitan/models/common/moe.py]: @tianyu-l would appreciate naming feedback on this lol

**Review by tianyu-l — COMMENTED** [review]
The change is a bit hard to read. My feeling was it's good enough if we just forbid dense SP + uneven seqlen in Decoder.post_init.

Ping me if you think something else is still necessary.

File: torchtitan/models/common/moe.py [hunk 2]
```diff
@@ -397,41 +399,46 @@ def forward(self, x_BLD: torch.Tensor) -> torch.Tensor:
...
         # all SP ranks, so physically pad to ``sp_size`` and trim before returning.
-        seq_pad = sp_size - L if L < sp_size else 0
-        if seq_pad:
-            if self.training:
-                raise ValueError(
-                    "MoE short-sequence padding for L < sp_size: "
-                    f"got L={L}, sp_size={sp_size} while the module is in training mode."
-                )
-            x_BLD = F.pad(x_BLD, (0, 0, 0, seq_pad))
-            L = L + seq_pad
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:417]: According to @wwwjn there might be an spmd_types / dtensor bug https://github.com/pytorch/torchtitan/issues/3622

Let's fix it sooner or later, but please don't delete this temporary TODO boundary in dotted lines.

If possible, please try to diagnose what happened. If too hard, leave for later.

File: torchtitan/models/common/moe.py [hunk 3]
```diff
@@ -397,41 +399,46 @@ def forward(self, x_BLD: torch.Tensor) -> torch.Tensor:
         operates on DTensors — the DTensor→local conversion happens at
         the GroupedExperts boundary.
         """
-        B, L, D = x_BLD.shape
-        sp_size = getattr(self.experts.token_dispatcher, "sp_size", 1)
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:402]: oh no, this code looks bad -- it says sp_size but it's wired to be TP mesh size, regardless of dense SP turned on/off

In general I think we should make `current_mesh_size` being able to tell SP on / off. Can leave as a future work.

File: torchtitan/models/common/moe.py [hunk 4]
```diff
@@ -397,41 +399,46 @@ def forward(self, x_BLD: torch.Tensor) -> torch.Tensor:
...
-                    "MoE short-sequence padding for L < sp_size: "
-                    f"got L={L}, sp_size={sp_size} while the module is in training mode."
-                )
-            x_BLD = F.pad(x_BLD, (0, 0, 0, seq_pad))
-            L = L + seq_pad
+        # Virtual padding then pads each batch's sequence length up to a multiple
+        # of ``sp_size`` without materializing padded tokens.
+        B, L, D = x_BLD.shape
+        sp_size = getattr(self.experts.token_dispatcher, "sp_size", 1)
+        if get_spmd_backend() == "spmd_types" and self.seq_sharded_input:
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: I feel the code should be for all backend.
The previous code sounds wrong to me when `seq_sharded_input=True`. Do you agree?

File: torchtitan/models/common/moe.py [hunk 5]
```diff
@@ -350,11 +350,13 @@ class Config(Module.Config):
         router: TokenChoiceTopKRouter.Config
         load_balance_coeff: float | None = 1e-3
         shared_experts: FeedForward.Config | None = None
+        seq_sharded_input: bool = False
 
     def __init__(self, config: Config):
         super().__init__()
 
         num_experts = config.num_experts
+        self.seq_sharded_input = config.seq_sharded_input
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: CP sounds independent, we mostly care if it's already sharded by TP -- it's also what you assign later.
```suggestion
        self.seq_dim_tp_sharded = config.seq_dim_tp_sharded
```

File: torchtitan/models/common/moe_sharding.py [hunk 6]
```diff
@@ -231,6 +231,7 @@ def set_moe_sharding_config(
     """
     # Always set sharding configs regardless of whether TP is enabled.
     # ``resolve_mesh`` filters out disabled axes at runtime.
+    moe_cfg.seq_sharded_input = enable_ep and enable_sp
```
Comment (tianyu-l) [inline torchtitan/models/common/moe_sharding.py]: This seems aligning with https://github.com/pytorch/torchtitan/blob/main/assets/images/moe_sharding.png -- only the topleft has sharded input into MoE.

@sanketpurandare and I were discussing improvement for these hacky logic. Let's work together later on them.

File: torchtitan/models/common/moe.py [hunk 7]
```diff
@@ -397,41 +399,46 @@ def forward(self, x_BLD: torch.Tensor) -> torch.Tensor:
...
+            # where CP/SP are off and local sequence length equals global
+            # sequence length. Compute the local TP stride from the unsplit
+            # MoE-region sequence length.
+            seq_pad = sp_size - L if L < sp_size else 0
+            if seq_pad:
+                if self.training:
+                    raise ValueError(
+                        "MoE short-sequence padding for L < sp_size: "
+                        f"got L={L}, sp_size={sp_size} while the module is in training mode."
+                    )
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: Let's raise in trainer post_init instead here, because CP is not considered.

Actually, it seems SP is already handled https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L152-L158
maybe add CP.
- seqlen should be divided by SP degree * [CP degree * 2] (if head-tail load balancer is used? not sure about other load balancer -- please check with @fegin)

File: torchtitan/models/common/moe.py [hunk 8]
```diff
@@ -350,11 +350,13 @@ class Config(Module.Config):
         router: TokenChoiceTopKRouter.Config
         load_balance_coeff: float | None = 1e-3
         shared_experts: FeedForward.Config | None = None
+        seq_sharded_input: bool = False
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: leave a TODO to kill this field later

File: torchtitan/models/common/moe_sharding.py (hunk 6 above)
Comment (sanketpurandare) [reply torchtitan/models/common/moe_sharding.py]: Yeah, I am also not too happy with the manual Partial construction after routed experts and using global indexing for slice insertion.

File: torchtitan/models/common/moe.py (hunk 4 above)
Comment (pianpwk) [reply torchtitan/models/common/moe.py]: yea the post-combine output was sp_size too small..

File: torchtitan/models/common/moe.py (hunk 3 above)
Comment (pianpwk) [reply torchtitan/models/common/moe.py:402]: SP is not a mesh setup property though, and it might be misleading inside/outside MoE regions, where SP-sharding varies ?

File: torchtitan/models/common/moe_sharding.py (hunk 6 above)
Comment (pianpwk) [reply torchtitan/models/common/moe_sharding.py]: TIL spmd_types does have Shard->Partial convert https://github.com/meta-pytorch/spmd_types/blob/645577f77c5abdb2b70d8f08468f08179e58f94e/spmd_types/_local.py#L952-L967

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/models/common/moe.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/models/common/moe.py:402]: We can have name like "MoE SP" (e.g. I believe vllm has that concept).

I believe today we are using `spmd_mesh_size` to determine how many shards we'd do on some mesh; in spirit, having SP degree available serves the same purpose.
